### PR TITLE
Fixed Templates and VM API

### PR DESCRIPTION
**Pull request by GilbertCherrie**
Fixes: https://github.com/ManageIQ/manageiq-api/issues/1117

Fixed the edit template API call and the VM controller "No Parent" option.

### CODE REVIEW

File: app/controllers/api/vms_controller.rb [hunk 1]
```diff
@@ -226,7 +227,11 @@ def build_parent_children(data)
                  end
 
       parent = if data.key?('parent_resource')
-                 fetch_relationship(data['parent_resource']['href'])
+                 if data['parent_resource'].nil?
+                  nil
+                 else
+                   fetch_relationship(data['parent_resource']['href'])
+                 end
```
Comment (kbrock) [inline app/controllers/api/vms_controller.rb:234]: think `dig` is your friend:

```ruby
  href = data.dig('parent_resource', 'href')
  parent = fetch_relationship(href) if href
```

File: app/controllers/api/templates_controller.rb [hunk 2]
```diff
@@ -4,8 +4,51 @@ class TemplatesController < BaseController
...
+
+      parent = if data.key?('parent_resource')
+                 if data['parent_resource'].nil?
+                   nil
+                 else
+                   fetch_relationship(data['parent_resource']['href'])
+                 end
+               end
+
+      [parent, Array(children)]
```
Comment (kbrock) [inline app/controllers/api/templates_controller.rb:44]: I think you want to use `Array.wrap()` here:

```suggestion
      [parent, Array.wrap(children)]
```

not sure if you want `Array(children)` or `Array.wrap(children)`
If `children = [child]`, did you want `[[child]]` or `[child]` coming back?
Also note `Array.wrap(nil) # => nil`

File: app/controllers/api/templates_controller.rb [hunk 3]
```diff
@@ -4,8 +4,51 @@ class TemplatesController < BaseController
...
+                   end
+                 end
+
+      parent = if data.key?('parent_resource')
+                 if data['parent_resource'].nil?
+                   nil
+                 else
+                   fetch_relationship(data['parent_resource']['href'])
+                 end
+               end
```
Comment (kbrock) [inline app/controllers/api/templates_controller.rb:42]:
```suggestion
            parent_href = data.dig('parent_resource', 'href')
            parent = fetch_relationship(parent_href) if parent_href
```

**Review by kbrock — COMMENTED** [review]
this looks good.

If you are just copy pasting, let me know and I can merge and then put my suggestions into another PR.

**Comment by GilbertCherrie** [comment]
> this looks good.
> 
> If you are just copy pasting, let me know and I can merge and then put my suggestions into another PR.

Yeah this was mostly copy paste. Not sure if this needs test cases or not and I want to finalize my UI PR before I take this off WIP.

**Comment by GilbertCherrie** [comment]
@miq-bot add-label bug

File: spec/requests/templates_spec.rb [hunk 4]
```diff
@@ -28,10 +28,48 @@
...
 
   context "editing a template" do
-    let!(:template) { FactoryBot.create(:template, :name => 'foo', :description => 'bar') }
+    let!(:template)     { FactoryBot.create(:template, :name => 'foo', :description => 'bar') }
+    let(:zone)          { FactoryBot.create(:zone, :name => "api_zone") }
+    let(:ems)           { FactoryBot.create(:ems_vmware, :zone => zone) }
+    let(:host)          { FactoryBot.create(:host) }
+    let(:template_openstack)       { FactoryBot.create(:template_openstack, :host => host, :ems_id => ems.id, :raw_power_state => "ACTIVE") }
+    let(:template_openstack1)      { FactoryBot.create(:template_openstack, :host => host, :ems_id => ems.id, :raw_power_state => "ACTIVE") }
+    let(:template_openstack2)      { FactoryBot.create(:template_openstack, :host => host, :ems_id => ems.id, :raw_power_state => "ACTIVE") }
```
Comment (kbrock) [inline spec/requests/templates_spec.rb:37]:
```suggestion
    let(:template_openstack)       { FactoryBot.create(:template_openstack, :host => host, :ems => ems, :raw_power_state => "ACTIVE") }
    let(:template_openstack1)      { FactoryBot.create(:template_openstack, :host => host, :ems => ems, :raw_power_state => "ACTIVE") }
    let(:template_openstack2)      { FactoryBot.create(:template_openstack, :host => host, :ems => ems, :raw_power_state => "ACTIVE") }
```

File: spec/requests/templates_spec.rb [hunk 5]
```diff
@@ -28,10 +28,48 @@
   end
 
   context "editing a template" do
-    let!(:template) { FactoryBot.create(:template, :name => 'foo', :description => 'bar') }
+    let!(:template)     { FactoryBot.create(:template, :name => 'foo', :description => 'bar') }
+    let(:zone)          { FactoryBot.create(:zone, :name => "api_zone") }
+    let(:ems)           { FactoryBot.create(:ems_vmware, :zone => zone) }
```
Comment (kbrock) [inline spec/requests/templates_spec.rb:33]: don't think zone is needed
```suggestion
    let(:ems)           { FactoryBot.create(:ems_vmware) }
```

if it is, then just ignore this suggestion

File: spec/requests/templates_spec.rb [hunk 6]
```diff
@@ -28,10 +28,48 @@
...
+    let(:host)          { FactoryBot.create(:host) }
+    let(:template_openstack)       { FactoryBot.create(:template_openstack, :host => host, :ems_id => ems.id, :raw_power_state => "ACTIVE") }
+    let(:template_openstack1)      { FactoryBot.create(:template_openstack, :host => host, :ems_id => ems.id, :raw_power_state => "ACTIVE") }
+    let(:template_openstack2)      { FactoryBot.create(:template_openstack, :host => host, :ems_id => ems.id, :raw_power_state => "ACTIVE") }
+    let(:new_vms) { FactoryBot.create_list(:template_openstack, 2) }
     before { api_basic_authorize(action_identifier(:templates, :edit)) }
     subject { send(req, api_template_url(nil, template), :params => params) }
 
+    describe "POST /api/templates/:c_id" do
+      it 'can edit a VM with an appropriate role' do
```
Comment (kbrock) [inline spec/requests/templates_spec.rb:43]:
```suggestion
      it 'can edit a template with an appropriate role' do
```

**Review by kbrock — COMMENTED** [review]
just those nits.

this looks great

**Comment by Fryguy** [comment]
If it's copy-paste, maybe it's better to make this a mixin named GenealogyMixin and included in both vms and templates.  Or we can do that in a follow up PR.

**Comment by GilbertCherrie** [comment]
> If it's copy-paste, maybe it's better to make this a mixin named GenealogyMixin and included in both vms and templates. Or we can do that in a follow up PR.

It is mostly copy paste. Do you want it as a mixix or is this PR good for now?

File: spec/requests/templates_spec.rb (hunk 4 above)
Comment (GilbertCherrie) [reply spec/requests/templates_spec.rb:37]: This change caused an error but I have implemented the other changes fine.

**Comment by kbrock** [comment]
Mixin sounds like a good route. Do you know how to do things like this?

**Comment by GilbertCherrie** [comment]
yeah I looked at some examples. I can give it a try.

**Comment by GilbertCherrie** [comment]
@Fryguy @kbrock I moved all the code to a mixin file. Is this good to merge now?

File: app/controllers/api/mixins/genealogy.rb [hunk 7]
```diff
@@ -0,0 +1,58 @@
+module Api
+  module Mixins
+    module Genealogy
+      RELATIONSHIP_COLLECTIONS = %w[vms templates].freeze
+      VALID_EDIT_ATTRS = %w[description name child_resources parent_resource].freeze
```
Comment (Fryguy) [inline app/controllers/api/mixins/genealogy.rb:5]: This is interesting...I'm wondering if we leave the valid attrs in the individual models and only "add" child_resources  and parent_resource...I'm ok with this being in a follow up though, since it might be fairly complicated?

File: app/controllers/api/mixins/genealogy.rb [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+      def tag_specified(id, data)
+        if id.to_i > 0
+          klass  = collection_class(:tags)
+          tagobj = klass.find(id)
+          return tag_path_to_spec(tagobj.name).merge(:id => tagobj.id)
+        end
+        parse_tag(data)
+      end
+
+      def assign_genealogy(type, id, data)
```
Comment (Fryguy) [inline app/controllers/api/mixins/genealogy.rb:16]: Not sure I like this name because it doesn't only assign genealogy...it also updates the rest of the attributes.  What about:

```suggestion
      def edit_resource_with_genealogy(type, id, data)
```

Comment (Fryguy) [reply app/controllers/api/mixins/genealogy.rb:16]: Alternatively we can just have `edit_resource` and call super, but I'm concerned that might get "lost" in the mixin and confuse people.  Curious what @kbrock thinks

File: app/controllers/api/mixins/genealogy.rb [hunk 9]
```diff
@@ -0,0 +1,58 @@
+module Api
+  module Mixins
+    module Genealogy
+      RELATIONSHIP_COLLECTIONS = %w[vms templates].freeze
+      VALID_EDIT_ATTRS = %w[description name child_resources parent_resource].freeze
+
+      def tag_specified(id, data)
```
Comment (Fryguy) [inline app/controllers/api/mixins/genealogy.rb:7]: I think all of these methods should be private so they are not exposed accidentally to the controller/router.

File: app/controllers/api/mixins/genealogy.rb (hunk 7 above)
Comment (Fryguy) [reply app/controllers/api/mixins/genealogy.rb:5]: I think the way to do this "right" is to have a VmsAndTemplates mixin.  In there we define `VALID_EDIT_ATTRS = %w[description name].freeze`, and remove it from here, and then also move `valid_custom_attrs` into VmsAndTemplate Mixin.  Finally, in the VmsAndTemplates mixin, you add in an include of Api::Mixins::Genealogy (alternatively just inline it right into the VmsAndTemplates mixin).

Then we add here: `VALID_GENEALOGY_ATTRS = %w[child_resources parent_resource].freeze`

Then down below we do 

```
        def validate_edit_data(data)
-         invalid_keys = data.keys - VALID_EDIT_ATTRS - valid_custom_attrs
+        invalid_keys = data.keys - VALID_EDIT_ATTRS - VALID_GENEALOGY_ATTRS - valid_custom_attrs
```

Something like that.

Comment (Fryguy) [reply app/controllers/api/mixins/genealogy.rb:5]: I can push a commit with my idea if that's what we want to do

File: app/controllers/api/mixins/genealogy.rb [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+                     end
+                   end
+
+        parent_href = data.dig('parent_resource', 'href')
+        parent = fetch_relationship(parent_href) if parent_href
+
+        [parent, Array.wrap(children)]
+      end
+
+      def fetch_relationship(href)
```
Comment (Fryguy) [inline app/controllers/api/mixins/genealogy.rb:50]: This might be too generically named now that it's in a mixin...not sure.

File: app/controllers/api/mixins/genealogy.rb (hunk 9 above)
Comment (Fryguy) [inline app/controllers/api/mixins/genealogy.rb:7]: Why is this method in this mixin?  Seems unrelated to genealogy

Comment (GilbertCherrie) [reply app/controllers/api/mixins/genealogy.rb:7]: Oh I copy and pasted this mixin template from another file, forgot to delete that.

File: app/controllers/api/mixins/genealogy.rb (hunk 7 above)
Comment (GilbertCherrie) [reply app/controllers/api/mixins/genealogy.rb:5]: I think that sounds good. If you want you can push a commit I can also try to do this on my end if you are busy.

Comment (GilbertCherrie) [reply app/controllers/api/mixins/genealogy.rb:5]: @Fryguy I pushed a commit with all the other suggestions, if you want I can also try implementing this one as well.

Comment (GilbertCherrie) [reply app/controllers/api/mixins/genealogy.rb:5]: @Fryguy I tried to implement this and it just broke my code. I'm assuming I did something wrong and not exactly sure which methods need to go in which mixin.

File: app/controllers/api/templates_controller.rb [hunk 11]
```diff
@@ -1,11 +1,16 @@
 module Api
   class TemplatesController < BaseController
+    include Api::Mixins::Genealogy
     include Subcollections::Policies
     include Subcollections::PolicyProfiles
     include Subcollections::Tags
 
-    def edit_resource(type, id = nil, data = {})
-      super(type, id, data.extract!('name', 'description'))
+    VALID_EDIT_ATTRS = %w[description custom_1 name child_resources parent_resource].freeze
```
Comment (Fryguy) [inline app/controllers/api/templates_controller.rb:8]: custom_1 should not be needed here because vm_custom_attrs will take care of all of the custom_# columns

**Review by Fryguy — APPROVED** [review]
LGTM - @kbrock can you also review?  We're going to defer some of the more complicated refactorings fro my other comment, because for now this works and moves @GilbertCherrie forward.  I'll be creating a follow up PR with the refactorings so we can discss that separately.

@kbrock, I'm not sure if you'd want more specs than this or if this is sufficient.

File: app/controllers/api/mixins/genealogy.rb (hunk 8 above)
Comment (kbrock) [reply app/controllers/api/mixins/genealogy.rb]: your name looks good. I also like the way Gilbert's new code reads.
Good suggestion

File: app/controllers/api/mixins/genealogy.rb (hunk 7 above)
Comment (kbrock) [reply app/controllers/api/mixins/genealogy.rb:5]: I do like the idea of having the decision of what keys are valid as part of the individual controller. nice call

File: app/controllers/api/mixins/genealogy.rb [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+      VALID_EDIT_ATTRS = %w[description name child_resources parent_resource].freeze
+
+      def edit_resource_with_genealogy(type, id, data)
+        attrs = validate_edit_data(data)
+        parent, children = build_parent_children(data)
+        resource_search(id, type).tap do |resource|
+          resource.replace_children(children)
+          resource.set_parent(parent)
+          resource.update!(attrs)
+        end
```
Comment (kbrock) [inline app/controllers/api/mixins/genealogy.rb:14]: in the followup, think this is the part that is genealogy. adding the children.
having the concept of valid params and the like seems part of the controller
having the concepts of adding children seems part of the mixin
